pkg/util/crypt/crypt_dev.go

```diff
@@ -0,0 +1,306 @@
+	stdin, err := cmd.StdinPipe()
+
+	if err != nil {
+		return "", "", err
+	}
+
+	go func() {
+		defer stdin.Close()
+		io.WriteString(stdin, input)
+	}()
+
+	err = cmd.Run()
+	if err != nil {
+		sylog.Verbosef("Unable to format crypt device")
+		return "", "", err
+	}
+
+	// Associate the temporary crypt file with a loop device
+	loop, err := createLoop(cryptF, 0, uint64(size))
+	if err != nil {
+		return "", "", err
+	}
+
+	sp := strings.Split(loop, "/")
+	loopdev := sp[len(sp)-1]
+
+	fd, err := lock.Exclusive("/dev/mapper")
+	if err != nil {
+		sylog.Debugf("Unable to acquire lock on /dev/mapper")
+		return "", "", err
+	}
+	nextCrypt := getNextAvailableCryptDevice()
+	cmd = exec.Command("/sbin/cryptsetup", "luksOpen", loopdev, nextCrypt)
+	cmd.Dir = "/dev"
+	stdin, err = cmd.StdinPipe()
+	if err != nil {
+		return "", "", err
+	}
+
+	go func() {
+		defer stdin.Close()
+		io.WriteString(stdin, input)
+	}()
+
+	err = cmd.Run()
+	if err != nil {
+		sylog.Verbosef("Unable to open crypt device: %s", nextCrypt)
+		return "", "", err
+	}
+
+	err = lock.Release(fd)
+	if err != nil {
+		sylog.Debugf("Unable to release lock on /dev/mapper")
+		return "", "", err
+	}
+
+	copyDeviceContents(path, "/dev/mapper/"+nextCrypt, squashFsSize)
+
+	// Open a new Temp file to stash the loop contents
+	loopF, err := ioutil.TempFile("", "loop-")
+	if err != nil {
+		sylog.Debugf("Error creating temporary crypt file")
+		return "", "", err
+	}
+
+	copyDeviceContents("/dev/"+loopdev, loopF.Name(), size)
+
+	return loopF.Name(), nextCrypt, err
+}
+
+// copyDeviceContents copies the contents of source to destination.
+// source and dest can either be a file or a block device
+func copyDeviceContents(source string, dest string, size int64) error {
+
+	squashFd, err := syscall.Open(source, syscall.O_RDWR, 0777)
+	if err != nil {
+		sylog.Debugf("Unable to open the file %s", source)
+		return err
+	}
+	defer syscall.Close(squashFd)
+
+	cryptFd, err := syscall.Open(dest, syscall.O_RDWR, 0777)
+
+	if err != nil {
+		sylog.Debugf("Unable to open the file: %s", dest)
+		return err
+	}
+
+	defer syscall.Close(cryptFd)
+
+	var writtenSoFar int64
+
+	buffer := make([]byte, 1024)
+	for writtenSoFar < size {
+		numRead, err := syscall.Read(squashFd, buffer)
+		if err != nil {
+			sylog.Debugf("Unable to read the the file %s", source)
+			return err
+		}
+		numWritten, err := syscall.Write(cryptFd, buffer)
+		if err != nil {
+			sylog.Debugf("Unable to write to destination %s", dest)
+			return err
+		}
+
+		if numRead != numWritten {
+			sylog.Debugf("numRead != numWritten")
+		}
+		writtenSoFar += int64(numWritten)
+	}
+
+	return nil
+}
+
+func getRandomString(n int) string {
+	b := make([]byte, n)
+	if _, err := rand.Read(b); err != nil {
+		sylog.Debugf("Something went wrong while generating random string")
+		return ""
+	}
+	return fmt.Sprintf("%x", b)
+}
+
+func getNextAvailableCryptDevice() string {
+	return getRandomString(15)
```

Comment:
or you can import `github.com/satori/go.uuid` and use `uuid.NewV4()`

Reply:
Used that instead